docs/docs/why-gatsby-uses-graphql.md

```diff
@@ -0,0 +1,362 @@
+  const products = require("./data/products.json")
+  products.forEach(product => {
+    createPage({
+      path: `/product/${product.slug}/`,
+      component: require.resolve("./src/templates/product.js"),
+      context: {
+        title: product.title,
+        description: product.description,
+        image: product.image,
+        price: product.price,
+      },
+    })
+  })
+}
+```
+
+The product template still uses `pageContext` to display the product data:
+
+```jsx:title=src/templates/product.js
+import React from "react"
+
+const Product = ({ pageContext }) => (
+  <div>
+    <h1>{pageContext.title}</h1>
+    <img
+      src={pageContext.image}
+      alt={pageContext.title}
+      style={{ float: "left", marginRight: "1rem", width: 150 }}
+    />
+    <p>{pageContext.price}</p>
+    <div dangerouslySetInnerHTML={{ __html: pageContext.description }} />
+  </div>
+)
+
+export default Product
+```
+
+Run `npm run develop`, then open `localhost:8000/product/space-socks/` to see one of the generated products.
+
+![Screenshot of a rendered product page.](./images/why-gql-product-json.png)
+
+This gets the job done, but it has a few shortcomings that are going to get more complicated as time goes on:
+
+1. The images and the product data are in different places in the source code.
+2. The image paths are absolute from the _built_ site, not the source code, which makes it confusing to know how to find them from the JSON.
+3. The images are unoptimized, and any optimization you do would have to be manual.
+4. To create a preview listing of all products, we’d need to pass _all_ of the product info in `context`, which will get unweildy as the number of products increases.
+5. It’s not very obvious where data is coming from in the templates that render the pages, so updating the data might be confusing later.
+
+To overcome these limitations, Gatsby introduces GraphQL as a data management layer.
+
+## Create pages using GraphQL
+
+There’s a bit more up-front setup required to get data into GraphQL, but the benefits far outweigh the cost.
+
+Using `data/products.json` as an example, by using GraphQL we’re able to solve all of the limitations from the previous section:
+
+1. The images can be colocated with the products in `data/images/`.
+2. Image paths in `data/products.json` can be relative to the JSON file.
+3. Gatsby can automatically optimize images for faster loading and better user experience.
+4. We no longer need to pass all product data into `context` when creating pages.
+5. Data is loaded using GraphQL in the components where it’s used, making it much easier to see where data comes from and how to change it.
+
+### Add the necessary plugins to load data into GraphQL
+
+In order to load the product and image data into GraphQL, we need to add a few [Gatsby plugins](/plugins/). Namely, we need plugins to:
+
+- Load the JSON file into GraphQL ([`gatsby-source-filesystem`](/packages/gatsby-source-filesystem/))
```

Comment:
I find the terminology a bit confusing here: what is a GraphQL node? What does it mean when data is loaded into GraphQL? Can we say something like "loaded into Gatsby's internal data store, that can be queried with GraphQL"?

Reply:
That makes sense, yeah. I'll clean it up.